paddle/framework/threadpool.h

```diff
@@ -67,7 +66,8 @@ class ThreadPool {
    * @return    std::future<void>, we could wait for the task finished by
    *            f.wait().
    */
-  std::future<void> Run(const Fun& fn) {
+  template <typename Callback>
```

Comment:
Here, use template will make `Run` method inline and remove an implicit casting.

It because if a user uses a lambda expression, every lambda is a different type. The lambda expression will cast to `std::function` first, then pass to `Run`. Use template will avoid the implicit cast.